### ISSUE TITLE
Add securityContext configuration to Helm Chart Values file, so runAsUser and runAsGroup can be set for coreDNS

### ISSUE THREAD

**Issue by GyunCho** [loft-sh/vcluster#742]
### Is your feature request related to a problem?

No

### Which solution do you suggest?

I may be misunderstanding how vcluster Helm Chart works, but currently, it seems it's not possible to set securityContext at both container and pod levels, so I can't set its runAsUser, runAsGroup, and fsGroup for coreDNS.

This is currently what I see when I <kubectl get pod coredns -o yaml ...>

```
securityContext:
  allowPrivilegeEscalation: false
  capabilities:
    drop:
    - all
  readOnlyRootFilesystem: true
```

...

securityContext: {}

and this is what I see from coreDNS' manifest file:

```
        securityContext:
          runAsUser: {{ .RUN_AS_USER }}
          runAsNonRoot: {{ .RUN_AS_NON_ROOT }}
          allowPrivilegeEscalation: false
          capabilities:
            drop:
            - all
          readOnlyRootFilesystem: true
```

...

```
      {{- if .Values.isolation.enabled }}
      securityContext:
        seccompProfile:
          type: RuntimeDefault
      {{- end }}
```

I'd like to set both runAsUser and runAsUser to 10001 for example through either Values.coredns.securityContext or Values.coredns.podSecurityContext.

[https://kubernetes.io/docs/tasks/configure-pod-container/security-context/](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/)

Thanks in advance!

### Which alternative solutions exist?

*No response*

### Additional context

*No response*

**Comment by rohantmp** [comment]
@GyunCho could you clarify the use-case?

The substitution for `runAsUser: {{ .RUN_AS_USER }}` is not done by helm. Instead the controller is the entity that deploys the coredns manifest and the `runAsUser` value is copied from its own UID.

It would be a small helm-only change to expose this in `values.yaml`, but I would like to understand the use-case better, and then I can send a PR :)

**Comment by GyunCho** [comment]
Hi @rohantmp!

This is to follow the security best practice.

"In the Kubernetes security context with runAsUser and runAsGroup. It is advisable to select a value greater than 10000, as this reduces the likelihood that this value is already taken on the host system."

Ref links: 
https://cloudogu.com/en/blog/k8s-app-ops-part-3-security-context-1
https://snyk.io/blog/10-kubernetes-security-context-settings-you-should-understand/

Thanks for explaining how ".RUN_AS_USER" works. I did not know how it works. 
So.. if I understood correctly, then I don't think there's a way for me to use it since I do not have any access to control the entity that deploy the coredns manifest. 

We use a CD tool that uses an agent, and modifying the UID of that CD tool's agent is not something I can do, or even if I can do, it's not encouraged to do.

Thanks!

**Comment by rohantmp** [comment]
Thanks for clarifying! @GyunCho 

**You can actually set these directly in** `Values.securityContext` as mentioned in this doc page: [Rootless mode & OpenShift].(https://www.vcluster.com/docs/operator/restricted-hosts)

**Those values will be copied to the coredns deployment.** 

When I mentioned the syncer controller, I meant the statefulset that runs the vcluster control-plane, and it is configured using `values.yaml`

Example `values.yaml`:
```yaml
securityContext:
  runAsUser: 12345
  runAsNonRoot: true
```

**Comment by rohantmp** [comment]
Does this cover your use-case?

**Comment by rohantmp** [comment]
Hi @GyunCho I just tried it out and it works for me. I'll post the details below. If the steps below don't work, then it might be a bug. In which case, please provide the vcluster version, the vcluster distributions (k3s(default),k0s, or k8s), values.yaml, and any other relevant details.


> I'm not so sure, but I believe what the doc really means is that by using "securityContext" under each component, you can set securityContext? not like set securityContext at .Values.securityContext and it's going to be set at the global level.

I can confirm now that it is a global level setting. The docs do mention: "The syncer also passes this UID down to the vcluster DNS deployment.."
Also, please note that the doc mentions that **rootless is only supported with the default k3s distribution** (others may work).

**Steps:**
Create values.yaml with:
```
securityContext:
  runAsUser: 12345
  runAsNonRoot: true
```
Create vcluster with ` vcluster create v1 -f values.yaml `
```

info   Creating namespace vcluster-v1
info   Detected local kubernetes cluster kind. Will deploy vcluster with a NodePort & sync real nodes
info   Create vcluster v1...
info   execute command: helm upgrade v1 https://charts.loft.sh/charts/vcluster-0.12.0.tgz --kubeconfig /tmp/1008074481 --namespace vcluster-v1 --install --repository-config='' --values /tmp/2024403050 --values values.yaml
done √ Successfully created virtual cluster v1 in namespace vcluster-v1
info   Waiting for vcluster to come up...
warn   vcluster is waiting, because vcluster pod v1-0 has status: ContainerCreating
warn   vcluster is waiting, because vcluster pod v1-0 has status: ContainerCreating
warn   vcluster is waiting, because vcluster pod v1-0 has status: ContainerCreating
info   Starting proxy container...
done √ Switched active kube context to vcluster_v1_vcluster-v1_kind-kind
- Use `vcluster disconnect` to return to your previous kube context
- Use `kubectl get namespaces` to access the vcluster

```
Now, coredns has those values in its secuirtyContext: 

In the vcluster: `kubectl get po -n kube-system -l k8s-app=kube-dns -o yaml |grep securityContext -A 7`
```
      securityContext:
        allowPrivilegeEscalation: false
        capabilities:
          drop:
          - ALL
        readOnlyRootFilesystem: true
        runAsNonRoot: true
        runAsUser: 12345
--
    securityContext: {}
    serviceAccount: coredns
    serviceAccountName: coredns
    terminationGracePeriodSeconds: 30
    tolerations:
    - effect: NoExecute
      key: node.kubernetes.io/not-ready
      operator: Exists
```


In the host cluster: `kubectl get po coredns-659f94b79f-wtlfj-x-kube-system-x-v1 -n vcluster-v1 -o yaml|grep securityContext -A 7`
```
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      readOnlyRootFilesystem: true
      runAsNonRoot: true
      runAsUser: 12345
--
  securityContext: {}
```

**Comment by rohantmp** [comment]
Closing for now. If it turns out to be a bug and the flow mentioned doesn't work, please reopen :)

**Comment by GyunCho** [comment]
Hi @rohantmp,

Sorry for not responding to you. I was on PTO in the past month.

Yes, I can confirm that the flow mentioned doesn't work. Maybe I'm terribly missing something really small, but I'm thinking it's probably because I'm using "EKS" Helm Chart instead of "k3s".

Also, I'd like to note that even if it works, it doesn't meet my requirements 100% as I also need to set runAsGroup in addition to runAsUser and runAsNonRoot.

Here are the environments that I've been using:
vcluster version: 0.12.0 and 0.12.2 

the vcluster distributions (k3s(default),k0s, or k8s): EKS

values.yaml:
      securityContext:
        runAsUser: 10001
        runAsNonRoot: true
        runAsGroup: 10001
      sync:
        ingresses:
          enabled: true
      syncer:
        extraArgs:
        - --override-hosts=false
        - --tls-san={{ load_balancer_domain_name }}
        - --out-kube-config-server=https://{{ load_balancer_domain_name }}
        resources:
          limits:
            cpu: 200m
            memory: 1Gi
          requests:
            cpu: 10m
            memory: 64Mi
        securityContext:
          allowPrivilegeEscalation: false
          runAsUser: 10001
          runAsGroup: 10001
          readOnlyRootFilesystem: true
          capabilities:
            drop:
              - ALL
      etcd:
        image: "eks-distro/etcd-io/etcd:{{ vcluster_images[kubernetes_version_for_tests | float].etcd.image_version }}"
        resources:
          limits:
            cpu: 200m
            memory: 1Gi
          requests:
            cpu: 20m
            memory: 150Mi
        fsGroup: 10001
        securityContext:
          allowPrivilegeEscalation: false
          runAsUser: 10001
          runAsGroup: 10001
          readOnlyRootFilesystem: true
          capabilities:
            drop:
              - ALL
      controller:
        image: "eks-distro/kubernetes/kube-controller-manager:{{ vcluster_images[kubernetes_version_for_tests | float].controller.image_version }}"
        resources:
          limits:
            cpu: 200m
            memory: 1Gi
          requests:
            cpu: 15m
            memory: 150Mi
        securityContext:
          allowPrivilegeEscalation: false
          runAsUser: 10001
          runAsGroup: 10001
          readOnlyRootFilesystem: true
          capabilities:
            drop:
              - ALL
      api:
        image: "eks-distro/kubernetes/kube-apiserver:{{ vcluster_images[kubernetes_version_for_tests | float].api.image_version }}"
        resources:
          limits:
            cpu: 1
            memory: 2Gi
          requests:
            cpu: 40m
            memory: 300Mi
        securityContext:
          allowPrivilegeEscalation: false
          runAsUser: 10001
          runAsGroup: 10001
          readOnlyRootFilesystem: true
          capabilities:
            drop:
              - ALL
      coredns:
        image: "eks-distro/coredns/coredns:{{ vcluster_images[kubernetes_version_for_tests | float].coredns.image_version }}"
        securityContext:
          runAsUser: 10001
          runAsGroup: 10001

**Comment by rohantmp** [comment]
Yes, looks like the EKS and K8S chart respect .Values.syncer.securityContext instead of .Values.securityContext, and in the EKS chart, the RUN_AS_USER is not escaped properly i.e. `{{ .RUN_AS_NON_ROOT }}` as opposed to ```{{`{{.RUN_AS_NON_ROOT}}`}}```.

**Comment by GyunCho** [comment]
Is there any update on this by any chance?
@rohantmp

**Comment by rohantmp** [comment]
Hi @GyunCho , will work on this this week :)

**Comment by eumel8** [comment]
Hi @rohantmp, vacation period is over, any progress here ? ;-)

**Comment by rohantmp** [comment]
In-progress now, just finishing up the PR!

**Comment by rohantmp** [comment]
Fixed for k3s coredns in #913 , but for EKS will need component.securityContext for example `syncer.securityContext` `api.securityContext`. I'll upload an example, but as mentioned early on, non-root mode is only guaranteed to work with k3s.

**Comment by rohantmp** [comment]
[Here](https://pastebin.com/raw/C3vS8NaU) is an example of the values used by k8s and eks. Initially I thought we should consume the same values, but on closer examination it looks like they do need a different values structure.
